mldsa/ntt.c

```diff
@@ -141,6 +141,44 @@ void ntt(int32_t a[MLDSA_N])
   /* are bounded in magnitude by 9 * MLDSA_Q                 */
 }
 
+/* Reference: Embedded into `invntt_tomont()` in the reference implementation
+ * [@REF] */
+static void mld_invntt_layer(int32_t r[MLDSA_N], unsigned layer)
+__contract__(
+  requires(memory_no_alias(r, sizeof(int32_t) * MLDSA_N))
+  requires(1 <= layer && layer <= 8)
+  requires(array_abs_bound(r, 0, MLDSA_N, (MLDSA_N >> layer) * MLDSA_Q))
+  assigns(memory_slice(r, sizeof(int32_t) * MLDSA_N))
+  ensures(array_abs_bound(r, 0, MLDSA_N, (MLDSA_N >> (layer - 1)) * MLDSA_Q)))
+{
+  unsigned start, k, len;
+  len = (MLDSA_N >> layer);
+  k = (1u << layer) - 1;
+  for (start = 0; start < MLDSA_N; start += 2 * len)
+  __loop__(
+    invariant(start <= MLDSA_N && k <= 255)
+    invariant(2 * len * k + start == 2 * MLDSA_N - 2 * len)
+    invariant(array_abs_bound(r, 0, start, (MLDSA_N >> (layer - 1)) * MLDSA_Q))
+    invariant(array_abs_bound(r, start, MLDSA_N, (MLDSA_N >> (layer - 1)) * MLDSA_Q)))
```

Comment:
This should be `(MLDSA_N >> layer)` not `(MLDSA >> (layer - 1))`

Reply:
but only in the second invariant...

Reply:
[6c53ce6](https://github.com/pq-code-package/mldsa-native/pull/179/commits/6c53ce6c53e2d6a6ebc4c95087f4ab28b0d0708a)